### ISSUE TITLE
Casts to ensure identical types?

### ISSUE THREAD

**Issue by vkuncak** [epfl-lara/stainless#994]
This code, for example,
```scala
  case class IntRef(var x:Int)
  def f = IntRef(42)

  def g(d1: IntRef, d2: IntRef) = {
     require(d2.x <= 100)
     val x1 : Int = d1.x
     d2.x = 33 + d2.x
  }
```
generates the checks at use sites that the variables have the right types. They seem a little confusing to a Scala programmer as they follow from declared types.
```scala
[ Debug  ] Symbols after AssertionInjector
[ Debug  ] 
[ Debug  ] -------------Functions-------------
[ Debug  ] def g(d1: IntRef, d2: IntRef): (Unit, IntRef) = {
[ Debug  ]   require({
[ Debug  ]     assert(d2 is IntRef, "Cast error")
[ Debug  ]     d2.x
[ Debug  ]   } <= 100)
[ Debug  ]   val d2: IntRef = d2
[ Debug  ]   val x1: Int = {
[ Debug  ]     assert(d1 is IntRef, "Cast error")
[ Debug  ]     d1.x
[ Debug  ]   }
[ Debug  ]   val d2: IntRef = IntRef({
[ Debug  ]     val rhs: Int = {
[ Debug  ]       assert(d2 is IntRef, "Cast error")
[ Debug  ]       d2.x
[ Debug  ]     }
[ Debug  ]     assert(((33 & -2147483648) == (rhs & -2147483648)) ==> ((33 & -2147483648) == (33 + rhs & -2147483648)), "Addition overflow")
[ Debug  ]     33 + rhs
[ Debug  ]   })
[ Debug  ]   val tmp: Unit = ()
[ Debug  ]   ((), d2)
[ Debug  ] }
```
Is it because we have refinements that we are not sure they hold? Should we special-case to avoid poluting the verification reports?

**Comment by vkuncak** [comment]
These casts are also generated when `IntRef` is marked as `final` class.

**Comment by vkuncak** [comment]
Another example: this code
```scala
  final case class IntRef(var x:Int)
  def g(d1: IntRef, d2: IntRef): Int = {
    val d3 = IntRef(d1.x + d2.x)
    d3.x
  }
```
gives this after assertion injector:
```scala
def g(d1: IntRef, d2: IntRef): Int = {
  val d3: IntRef = IntRef({
    assert(d1 is IntRef, "Cast error")
    d1.x
  } + {
    assert(d2 is IntRef, "Cast error")
    d2.x
  })
  assert(d3 is IntRef, "Cast error")
  d3.x
}
```
Why all these casts, @samarion and @jad-hamza ?

**Comment by jad-hamza** [comment]
I think they come from here, we can avoid generating the assertion when there is just one constructor:

https://github.com/epfl-lara/stainless/blob/940261c8669beba76cff59fc05eea2ddfd4a6988/core/src/main/scala/stainless/verification/AssertionInjector.scala#L79-L85

**Comment by samarion** [comment]
We can consider adding an `Unchecked` annotation when the `ADTSelector` is created in the AdtSpecialization phase otherwise: https://github.com/epfl-lara/stainless/blob/940261c8669beba76cff59fc05eea2ddfd4a6988/core/src/main/scala/stainless/extraction/oo/AdtSpecialization.scala#L73

**Comment by jad-hamza** [comment]
If we add `Unchecked` the type-checker will also ignore assertions that appear in `expr` (I guess we have this problem in other places where we use `Unchecked`)

**Comment by samarion** [comment]
You can just add a let-binding to circumvent this.

**Comment by jad-hamza** [comment]
At the moment let bindings don't work well with `SizedADT` because `getType` returns `ADT`

**Comment by samarion** [comment]
Should we maybe add a new annotation which only ignores directly nested assertions/checks (instead of all checks within the annotated expression)?

**Comment by samarion** [comment]
Or rather, shouldn't we fix the `SizedADT` issue instead of introducing a bunch of extra hacks to work around it?

**Comment by jad-hamza** [comment]
Yes that'd be great, but these are big changes, what do you have in mind?
Should `SizedADT` be moved to Inox?

**Comment by samarion** [comment]
Can't we fix this by adding an untyped let-binding? Somethere where the let-binding type isn't checked but rather inferred. We can control this through a flag on the `ValDef` or add a type annotation.

**Comment by jad-hamza** [comment]
This is fixed now by I'm reopening for the untyped let-binding and the `Unchecked` annotation.
